pkg/reconciler/ingress/status.go

```diff
@@ -445,7 +448,8 @@ func (m *StatusProber) listGatewayURLsPerPods(gateway *v1alpha3.Gateway) (map[st
 				}
 
 				for _, addr := range sub.Addresses {
-					urlsPerPods[addr.IP] = append(urlsPerPods[addr.IP], fmt.Sprintf(urlTmpl, portNumber))
+					ingressPodAddress := net.JoinHostPort(addr.IP, strconv.Itoa(int(portNumber)))
```

Comment:
This won't work. If you look where it is later used, all the URLs of a specific map entry (i.e. pod IP) are handled together and once succeeds the pod is considered probed. Here, by introducing `portNumber` in the map key, you are breaking the entire model.

I would suggest that instead of returning a string per entry, your return a struct containing:
* `urlTmpl`: `fmt.Sprintf(urlTmpl, int32(server.Port.Number))`
* `targetPort`: `portNumber`

Store `targetPort` in `WorkItem` and use it in `DialContext`.